### PR TITLE
ARROW-5181: [Rust] Initial support for Arrow File reader

### PR DESCRIPTION
**Pull request by nevi-me**
This adds initial support for reading Arrow files. Only the file format is supported, with support for the streaming format to follow in a separate PR.

Only Rust supported datatypes are read and tested, thus files with timestamp[tz], intervals, durations, maps, decimal; aren't tested.

### CODE REVIEW

**Comment by nevi-me** [comment]
Hi @paddyhoran @andygrove @sunchao please review when you get a chance. We can place the reusable parts (validating headers, reading schemas and batches) in a common module when we work on the streaming format.

**Comment by agrove-rms** [comment]
@nevi-me I'm excited to see this! I will make time to review over the weekend.

File: rust/arrow/src/ipc/file/reader.rs [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+                        block.offset() as u64 + block.metaDataLength() as u64,
+                    ))?;
+                    self.reader.read_exact(&mut a_data)?;
+
+                    // construct buffers from their blocks
+                    let c_buffers = c_batch.buffers().unwrap();
+
+                    // get fields and determine number of buffers to use for each
+                    let c_nodes = c_batch.nodes().unwrap();
+                    let mut buffer_num = 0;
```
Comment (sunchao) [inline rust/arrow/src/ipc/file/reader.rs]: nit: can we name these `buffer_index` and `field_index`?

File: rust/arrow/src/ipc/file/reader.rs [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+                        field_num = field_num + 1;
+                        buffer_num = buffer;
+
+                        arrays.push(array);
+                    }
+
+                    RecordBatch::try_new(self.schema.clone(), arrays)
+                        .map(|batch| Some(batch))
+                }
+                ipc::MessageHeader::SparseTensor => panic!(),
```
Comment (sunchao) [inline rust/arrow/src/ipc/file/reader.rs]: nit: can we just have a default branch with `unimplemented` error?

File: rust/arrow/src/ipc/file/reader.rs [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Read a specific record batch
+    ///
+    /// Sets the current block to the batch number, and reads the record batch at that
+    /// block
+    pub fn read_batch(&mut self, batch_num: usize) -> Result<Option<RecordBatch>> {
```
Comment (sunchao) [inline rust/arrow/src/ipc/file/reader.rs]: nit: wondering if we can change this to something like `set_index` which just change the `current_block` and the caller then have to call `next()` to actually read the batch.

Otherwise, this does have a side effect of advancing the current block index, which may caught people by surprise. We should at least point this out in the method comments.

File: rust/arrow/src/ipc/convert.rs [hunk 4]
```diff
@@ -61,6 +63,108 @@ fn schema_to_fb(schema: &Schema) -> FlatBufferBuilder {
...
+                (32, ipc::TimeUnit::SECOND) => DataType::Time32(TimeUnit::Second),
+                (32, ipc::TimeUnit::MILLISECOND) => {
+                    DataType::Time32(TimeUnit::Millisecond)
+                }
+                (64, ipc::TimeUnit::MICROSECOND) => {
+                    DataType::Time64(TimeUnit::Microsecond)
+                }
+                (64, ipc::TimeUnit::NANOSECOND) => DataType::Time64(TimeUnit::Nanosecond),
+                z @ _ => panic!(
+                    "Time type with bit witdh of {} and unit of {:?} not supported",
```
Comment (sunchao) [inline rust/arrow/src/ipc/convert.rs]: witdh -> width

**Review by sunchao — COMMENTED** [review]
Sorry for the late review @nevi-me . Overall looks good.

**Comment by nevi-me** [comment]
Hi @sunchao I'm currently working on this PR to add support for lists and structs, as I've now generated the data.

File: rust/arrow/src/ipc/file/reader.rs [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+use crate::buffer::Buffer;
+use crate::datatypes::{DataType, Schema};
+use crate::error::{ArrowError, Result};
+use crate::ipc;
+use crate::record_batch::RecordBatch;
+
+static ARROW_MAGIC: [u8; 6] = [b'A', b'R', b'R', b'O', b'W', b'1'];
+
+/// Read a buffer based on offset and length
+fn read_buffer(c_buf: &ipc::Buffer, a_data: &Vec<u8>) -> Buffer {
```
[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/arrow/src/ipc/file/reader.rs]: I'll change the naming, I was using `c_` to remember that I'm dealing with converted data from flatbuffers

**Comment by sunchao** [comment]
Cool. Looking forward to the updated PR!

File: rust/arrow/src/ipc/convert.rs [hunk 6]
```diff
@@ -17,14 +17,16 @@
...
-use crate::datatypes::DataType::*;
-use crate::datatypes::Schema;
+use crate::datatypes::{DataType, DateUnit, Field, Schema, TimeUnit};
 use crate::ipc;
 
 use flatbuffers::FlatBufferBuilder;
 
 /// Serialize a schema in IPC format
+///
+/// TODO(Neville) add bit-widths and other field metadata to flatbuffer Type
```
[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/arrow/src/ipc/convert.rs]: Removed from doc

File: rust/arrow/src/ipc/file/reader.rs [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        assert!(batch.is_none());
+    }
+
+    fn validate_batch(batch: RecordBatch) {
+        assert_eq!(5, batch.num_rows());
+        assert_eq!(4, batch.num_columns());
+        let arr_1 = batch.column(0);
+        let int32_array = Int32Array::from(arr_1.data());
+        assert_eq!(
+            "PrimitiveArray<Int32>\n[\n  1,\n  2,\n  3,\n  null,\n  5,\n]",
```
[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/arrow/src/ipc/file/reader.rs]: I've come across a problem here, as a result the current commit will (intentionally) fail. Please see my comment on the PR below

**Comment by nevi-me** [comment]
Hi @sunchao I've updated the PR with list and struct reading. Regarding your comment about using `ArrayData`, I've noticed that there's a difference between Rust and Python/CPP (maybe just IPC) with buffer padding to multiples of 8 (and treatment of null buffers, but I'll address this separately).

The current commit ab6ff9b (https://github.com/apache/arrow/pull/4167/commits/ab6ff9bfa930550df09a5f575b8724478045d176#diff-cd3519a5e748548b5323b0d05d8e9c8aR504) will fail with the below:

```rust
thread 'ipc::file::reader::tests::test_read_struct_file' panicked at 'assertion failed: `(left == right)`
  left: `ArrayData { data_type: Boolean, len: 5, null_count: 2, offset: 0, buffers: [Buffer { data: BufferData { ptr: 0x20a62bba200, len: 8 }, offset: 0 }], child_data: [], null_bitmap: Some(Bitmap { bits: Buffer { data: BufferData { ptr: 0x20a62bb9fc0, len: 8 }, offset: 0 } }) }`,
 right: `ArrayData { data_type: Boolean, len: 5, null_count: 2, offset: 0, buffers: [Buffer { data: BufferData { ptr: 0x20a62bbec00, len: 1 }, offset: 0 }], child_data: [], null_bitmap: Some(Bitmap { bits: Buffer { data: BufferData { ptr: 0x20a62bbf000, len: 1 }, offset: 0 } }) }`', arrow\src\ipc\file\reader.rs:510:13
```

The difference being the `len: 1` vs `len: 8` in the `BufferData` lengths. The above is from a `BooleanArray` with 5 elements. I've run out of time for now, but I'll compare the binary rep of the buffers to see what's causing the difference, but I suspect that it's just the lengths being padded. https://www.diffchecker.com/xU8bBtne shows the diff between the `struct_array` buffer and the `struct_type` file that I'm reading. The padding to multiples of 8 is more apparent there.

There's other differences that I've picked up as I was going along, which might affect our interop compatibility, but I'll list/address them over the coming days.
I'll also add the python scripts that I used to generate the data, but I'd prefer to move the data files that I'll use to `arrow-testing` during the course of this PR.

File: rust/arrow/src/ipc/file/reader.rs [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+    pub fn num_batches(&self) -> usize {
+        self.total_blocks
+    }
+
+    /// Return the schema of the file
+    pub fn schema(&self) -> Arc<Schema> {
+        self.schema.clone()
+    }
+
+    /// Read the next record batch
```
Comment (nevi-me) [inline rust/arrow/src/ipc/file/reader.rs]: Will this remain appropriate as returning `RecordBatch`, given that we can return dictionaries and tensors in the future?

**Comment by andygrove** [comment]
@nevi-me This is looking great . I think this change is big enough that we should update the README too and explain where those test data files came from and how there were created.

**Comment by sunchao** [comment]
> Update: The buffers from the Arrow file are padded to 64 bits, while the ones in Rust are padded to 8-bits. Due to this difference, I can't test data equality using ArrayData.

In Rust we also pad buffer with 64 bits. I think the real reason here is that the `len` in `BufferData` is the **# of valid bytes**, instead of the # of total bytes (= # of valid bytes + # of padded bytes). I'm not sure whether it is easy to change the conversion of `ipc::Buffer` to meet this requirement. Changing the `BufferData` implementation might be a little involving but let me take a look.

File: rust/arrow/src/ipc/file/reader.rs (hunk 8 above)
Comment (sunchao) [reply rust/arrow/src/ipc/file/reader.rs]: Good point. For dictionary it is OK as we use the dictionary to decode record batches. We may need additional methods to read tensor/sparse tensor. We can use the [C++ implementation](https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/reader.h) as a reference.

**Comment by nevi-me** [comment]
Thanks @sunchao

File: rust/arrow/src/array.rs [hunk 9]
```diff
@@ -1034,6 +1034,35 @@ impl From<Vec<(Field, ArrayRef)>> for StructArray {
     }
 }
 
+impl From<(Vec<(Field, ArrayRef)>, Buffer, usize)> for StructArray {
+    fn from(triple: (Vec<(Field, ArrayRef)>, Buffer, usize)) -> Self {
+        let (field_types, field_values): (Vec<_>, Vec<_>) = triple.0.into_iter().unzip();
+
+        // Check the length of the child arrays
+        let length = field_values[0].len();
```
Comment (pitrou) [inline rust/arrow/src/array.rs]: Just a drive-by comment, but what happens if the vector is empty?

Comment (nevi-me) [reply rust/arrow/src/array.rs]: I'd personally prefer panicking in this instance. I'd think it reasonable for the caller of the conversion method(s) to ensure they don't pass empty lists. We can at least `panic!("Field value list should not be empty when constructing struct array")`. We could use `TryFrom`, but I feel it would be overkill.

Comment (pitrou) [reply rust/arrow/src/array.rs]: That said, for the record, the C++ side allows 0-field struct arrays (even though they don't seem tested).

Comment (nevi-me) [reply rust/arrow/src/array.rs]: Perhaps if an empty record batch and empty schema are fine (which I think both are), then we can/must allow there to be an empty struct array in general.
The conversion method in question might still need to panic, but I'll give this more thought.

File: rust/arrow/src/ipc/convert.rs [hunk 10]
```diff
@@ -61,6 +63,108 @@ fn schema_to_fb(schema: &Schema) -> FlatBufferBuilder {
     fbb
 }
 
+/// Deserialize a Schema table from IPC format to Schema data type
+pub fn fb_to_schema(fb: ipc::Schema) -> Schema {
```
[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/arrow/src/ipc/convert.rs]: I've added one test to rule them all, by converting a large `Schema` with all the types that we support, to flatbuffers and back. Would that suffice?

**Comment by wesm** [comment]
What is the status of this patch for 0.14.0?

**Comment by nevi-me** [comment]
> What is the status of this patch for 0.14.0?

Hi @wesm, this might not make 0.14. I ended up being dependent on a few other changes that we needed to make. I'll be travelling next week, and doubt I'll be able to complete this in time for 0.14.

**Comment by wesm** [comment]
OK, no problem, I removed from the milestone

**Comment by jbabyhacker** [comment]
@nevi-me & @wesm what is the status of this PR?

**Comment by nevi-me** [comment]
Hi @jbabyhacker 

**TL;DR** it's not abandoned.

The long version's that I haven't had enough bandwidth to work on it. I got stuck because I needed help with some changes, but by the time we had introduced those changes, I was already swamped. I've been working long hours at work for the past few months, including weekends. I'm nearly done with the current project at work, so I'm anticipating having downtime from next weekend,. So my time will be broken down between my studies and catching up here.

In terms of work required, I still need to:

- add test files to `arrow-testing` https://issues.apache.org/jira/browse/ARROW- 5399
- update test cases after we merge the above, including adding more tests for coverage purposes

I intend on getting IPC (reader, writer, stream vs batch) before 1.0.0.

**Comment by lihalite** [comment]
If it's interesting at all, I put together a minimal Arrow Flight for Rust proof-of-concept on top of this branch: https://github.com/lihalite/arrow/commit/5ace5b226fb4a3a2a445b11c5b13f847ee3991b1

I used tower-grpc. The impl in this CR is enough to deserialize Flight schema and record batch messages, so we can make a ListFlights call followed by a DoGet.

**Comment by wesm** [comment]
That's cool. It seems like hardening IPC in Rust is a pre-requisite for many other projects. It might need to be a collective effort rather than blocking on one person

**Comment by lihalite** [comment]
Maybe we can kick off a discussion on the mailing list?
- Status of Rust IPC, other subprojects there. Maybe it's reasonable to merge this as a start and build out more test cases and the other APIs?
- Flight support in Rust: which gRPC implementation, which Protobuf implementation, etc. (tower-grpc isn't quite up to par with the standard gRPC implementations, but the others bind to C or C++ IIRC)

**Comment by nevi-me** [comment]
My apologies for dragging so much on this, I've been on one of those "it might end next week" projects that's taken too much of my time.

In terms of effort, the reader is complete, supporting all types that Rust can read. I'm mainly left with semantics and moving test cases to `arrow-testing`. I'll work on wrapping this up by the weekend so we can unlock progress here.

I haven't been able to follow the discussions around buffer alignment, so I might need help there from someone else.

As a start, I'll rebase when I get home tonight.

File: rust/arrow/src/ipc/file/reader.rs (hunk 3 above)
Comment (nevi-me) [reply rust/arrow/src/ipc/file/reader.rs]: Done

**Comment by nevi-me** [comment]
# Status of this PR

The file reader is done, however unit tests are failing because our Rust impl doesn't yet support a number of types:

- timestamp[tz] > the schema comparisons fail as we do not keep timezone metadata
- binary > we only support utf8
- fixedsizelist > not supported
- large{utf8|list} > not supported

I've already spent a lot of time on being able to use the integration testing files, so I'm opposed to my initial idea of creating a subset of files that only have Rust-supported types (unless there's a more automated way of doing it).

## Proposals

1. We disable the test, and merge this PR.
2. We keep this PR open, and revisit after working on some of the above issues.
3. We generate integration files for Rust excluding the above types (if there's a more automated process than what I was manually doing).

Any suggestions @andygrove @wesm @emkornfield?

**Comment by andygrove** [comment]
@nevi-me What is the nature of the current test failures? Does the code panic or return an `Err`? If the code returns an `Err` stating that some type is not supported then I would have the test pass if the current expected error is returned.

**Comment by andygrove** [comment]
> I'm opposed to my initial idea of creating a subset of files that only have Rust-supported types (unless there's a more automated way of doing it).

I think we can automate this, by calling the Java implementation to generate some files for us and we can do all of this with docker. I think I'll start exploring this.

**Comment by nevi-me** [comment]
> > I'm opposed to my initial idea of creating a subset of files that only have Rust-supported types (unless there's a more automated way of doing it).
> 
> I think we can automate this, by calling the Java implementation to generate some files for us and we can do all of this with docker. I think I'll start exploring this.

On the status of the error, it panics on unsupported datatypes. I think the CI failures happen before it gets to that stage, so I'll look into that.

I'm currently working on `FixedSizeListArray`, I'll get the minimum viable support for it, so I can see if the reading of files that have this type work. Separate files that only include supported types will help, so thank you.

**Comment by nevi-me** [comment]
> I'm currently working on `FixedSizeListArray`, I'll get the minimum viable support for it, so I can see if the reading of files that have this type work. Separate files that only include supported types will help, so thank you.

FixedSizeListArray is done and working, I've also worked on splitting BinaryArray with StringArray, as well as FixedSizeBinaryArray. I'll submit PRs for them soon, Need to write more tests for them and clean up their code.

On my local branch, I can read the integration files, with one odd exception where f64 arrays are being read at the incorrect record length.

**Comment by andygrove** [comment]
Great progress @nevi-me !

I'm traveling this week but will start helping with integration testing next week.

**Comment by nevi-me** [comment]
@paddyhoran @andygrove @sunchao @liurenjie1024 PTAL. If anyone has capacity, it would be great to add more data type support so we can be able to read more integration files.

I'm going to work on the stream format reader next, which will be very useful for `flight`.

**Review by andygrove — APPROVED** [review]
@nevi-me I don't pretend to understand every detail here but LGTM overall. There are a couple of unwrap/panics that could probably be changed to use results but I think we should get this PR merged to  make it easier for other committers (myself included) to start helping you out with this effort.

**Review by paddyhoran — APPROVED** [review]
I agree with @andygrove.  Let's get this merged so that others can start to chip in.  Thanks @nevi-me, this is great progress!

**Comment by andy-thomason** [comment]
What is the status of this? I can only see the beginnings of a schema converter in the repo
and went ahead and wrote a full version of this. I also have batch and dictionary support in
internal work projects if this is required but would be overjoyed if someone had done this already.

https://github.com/andy-thomason/arrow

**Comment by nevi-me** [comment]
Hi @andy-thomason, your `master` branch is 725 commits behind `apache/arrow`. It's often a good idea to rebase before doing any work 😃
The reader was merged 5 days ago, so you can review that and perhaps continue from there. dictionary support is exciting to me! Its relevant JIRA is https://issues.apache.org/jira/browse/ARROW-5949, so perhaps you can open a PR against it as a start.

I haven't had time to work on the stream reader (as this PR was only for the file reader), and we'd welcome more hands on deck so we can complete IPC in Rust by `1.0.0` (not sure of when that'll be, but might be in December)

**Comment by andy-thomason** [comment]
Excellent news, @nevi-me We use arrow extensively at work for genetic variant analysis and I have written a number of parsers for the format - all internal, I'm afraid. I'll sync up and take a look.

**Comment by nevi-me** [comment]
Is this all using Rust @andy-thomason?

**Comment by andy-thomason** [comment]
We were using C++, Python and R in the past, but now we are starting to write services in Rust instead of running batch code on a cluster.

**Comment by andy-thomason** [comment]
We also have a group in Cambridge using Rust in WASM.

**Comment by andygrove** [comment]
This related work also just got merged.

https://github.com/apache/arrow/tree/master/rust/arrow-flight

**Comment by andy-thomason** [comment]
Excellent @andygrove, Saved me a huge pile of work. I have a deadline of the end of next week to
get a file reader working using hyper. Liking this more and more.

**Comment by nevi-me** [comment]
If you'll be using `arrow-flight`, the stream reader might be the next best thing for us to work on. I can't remember the spec, but most of the work should be refactoring the file reader to check for end of stream

**Comment by andy-thomason** [comment]
I'll certainly look at `arrow-flight`. At the moment we are using static data servers as our data is significantly read only, but I would like to move to a microservice model. We use record batches of about 1GB and have a second arrow file as a sparse index.

**Comment by nevi-me** [comment]
JIRA logs 10 minutes to each PR for comments, so maybe the mailing list might be a better place to continue the conversation (as this PR is closed). Be wary though that gRPC has low data limits by default (4MB, haven't tried modifying it in `tonic` or `tower-grpc`, but you can change it with a setting in other languages). You might end up having to break your arrow batches down to smaller chunks.
